### ISSUE TITLE
Could new signature args be added to an event?

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#77]
From https://github.com/WordPoints/hooks-api/issues/59#issuecomment-167158832:

> The only thing we need to consider is events with multiple args. Since there really is no concept of "main arg" in the API, do we consider a fire a repeat only if all of the args are the same? I guess that makes sense. But the question then is, what about when args are added or removed from the event? The fires will then have a different "signature", even if the remaining args are the same. This is an issue that came up when considering handling reversals in the periods API (something we decided not to do at the moment). The proposed solution was to store the values of all of the (top level) event args and then only compare the args in the current fire that were also in the past fire. Then we'd consider it a repeat only if all of them are the same.

The difference between this and the periods API is that in there we are dealing with stateful args in addition to signature args. Here we are dealing only with signature args. This leads me to question whether there would ever be any good reason to add/remove _signature_ args. That would be changing the very nature of the event, since each signature arg is modified by the event. But what if a plugin comes along and adds a new entity that participates in the event and is somehow modified by it? But then shouldn't that be a different event? And if not, wouldn't the new arg likely be directly related to one of the existing args, so that adding it as another arg would be unnecessary?

Let's look at this from another angle. If signature args can be added to and removed from events at will, non-repeatable events could even become repeatable, and vice versa. In other words, the nature of the event changes, like we said above. This makes me question even more the wisdom of supporting this.

Let's try to answer that question we asked above: would it really ever make sense for a new entity to participate in an event such that it is modified, when that entity is not related to the other entities in the event in any way? That is difficult to imagine, because every scenario that I can think of would really only make sense if the entity was related to one of the other involved entities in some way. And in that case, what we would do is add a new entity relationship, not a new event arg. When I try to imagine a case where the new entity wasn't related, I immediately think of stateful args, which are essentially just "bystanders".

**This just really seems impossible, but I'm not sure that we can prove that it is. What we can say is this: if an entity takes part in the event but is not related to the other entities, what we really have is a "parallel" event, and a new event should be registered for it, instead of adding an arg to an existing event.**

I think it would be helpful to document here what kind of behavior would result if a signature arg was added/removed. Also, we might want to consider the possibility that an event could be registered without all of the args that it could have been registered with, leading us to want to add an arg later. Or, what if the code behind the event itself were to change so that an arg was no longer modified in the event anymore, or no longer involved at all?

**Comment by JDGrimes** [comment]
## Side affects of adding a signature arg

### Non-repeatable events

First let's consider the case when there are already some other signature args, and so the event is non-repeatable. Adding another signature arg will expose more options in the reactions UI. It will also affect behavior in identifying repeat fires. Any hit which occurred with the previous set of signature args would be ignored when checking for repeats (based on the current database structure), with the result that each hit that had already occurred would be able to repeat a single time.

### Repeatable events

If the event is repeatable, there are no signature args until the new one is added. Adding it causes the event to now become non-repeatable, meaning the behavior will instantly change. Any prior hits will not be ignored when detecting repeats. Repeat detection will work properly from that point on though. The UI may also change, since the periods extension bases its exposure in the UI on whether an event is repeatable or not. Any configured periods will then become hidden.

## Side affects of removing a signature arg

Since repeatable events don't have any signature args to start with, we're only talking about non-repeatable events here. Of course, if there is only a single signature arg for an event, and it is being removed, the event will become repeatable.

When that happens any prior hits will be ignored since no repeat detection will be attempted. Also, the UI may change in response to this as well, but currently it would only result in more options being available, not in any current options being hidden.

On the other hand, when there are multiple signature args and so the event remains repeatable, all that will happen is that prior hits will be ignored when detecting repeats.

In both cases, since an arg is being removed, existing reactions may cease to function anyway, if they reference that arg at all. The UI for those reactions may not function correctly for the same reason.

**Comment by JDGrimes** [comment]
> Also, we might want to consider the possibility that an event could be registered without all of the args that it could have been registered with, leading us to want to add an arg later. Or, what if the code behind the event itself were to change so that an arg was no longer modified in the event anymore, or no longer involved at all?

From the discussion in the above comment we can glean some information that will be helpful in answering these questions.

First, removing an arg entirely is something that is going to cause a lot of pain anyway. There's really no elegant way to do that, and so it should be avoided at all cost.

That isn't to say that a _signature_ arg could never be "removed"; it could simply become a stateful arg instead, which would avoid those issues. However, even in that case, we wouldn't ever want to do this when it would cause the event to become repeatable. That would be changing the very nature of the event itself, and would be retroactively changing the behavior of the existing reactions. There is no elegant way to handle that situation, the best thing to do is to disable the reactions, or ideally, to continue handling them just as before (unless this wasn't the expected behavior, in which case we'd just be fixing a bug).

This means that we'll have much bigger fish to fry than worrying about bugs with repeat detection, _except in the case where the event began with multiple signature args, and so remains repeatable._ I have to admit that I'm actually not convinced that events could or should have multiple signature args, but that is a discussion, if not for another time, at least for another ticket. Assuming such a thing could exist, what would happen if one of the signature args was removed is that prior hits will not be taken into account when detecting repeats. So our worst case scenario is that some users get awarded points twice instead of just once. Not ideal, sure, but not the end of the world either. That also just happens to be the worst case scenario when adding a new signature arg (ignoring the possibility that the event was previously repeatable, since that carries other problems with it, and actually is less severe proportionately speaking—if the event was non-repeatable to start with, the user could game 100%, or double the points that they should, out of the system: if the event was repeatable before, that approaches a minimum of 0%, since they could have been awarded many, many times before).

**Comment by JDGrimes** [comment]
The discussion from #78 was very useful. There we concluded that each event can only really have one primary or "signature" arg. Therefore we can limit the possible cases that we need to consider in this ticket. We no longer need to think about the cases where there are multiple "signature" args, because that is an impossibility. This means that there are only two possibilities that we need to consider:
- When a primary arg is added to a repeatable event.
- When the primary arg is removed from a non-repeatable event.

In both of these cases, as we pointed out above, the very nature of the event is changed from repeatable to non-repeatable, or vice versa.

The results of this (from above):

> If the event is repeatable, there are no signature args until the new one is added. Adding it causes the event to now become non-repeatable, meaning the behavior will instantly change. Any prior hits will not be ignored when detecting repeats. Repeat detection will work properly from that point on though. The UI may also change, since the periods extension bases its exposure in the UI on whether an event is repeatable or not. Any configured periods will then become hidden.
> 
> [I]f there is only a single signature arg for an event, and it is being removed, the event will become repeatable.
> 
> When that happens any prior hits will be ignored since no repeat detection will be attempted. Also, the UI may change in response to this as well, but currently it would only result in more options being available, not in any current options being hidden.

It seems to me that the worst case scenario, as discussed above, is when removing the primary arg and causing the event to become repeatable, since that will cause the event to begin firing willy-nilly, as no periods will be in place to restrict it. The second worst thing that happens is when adding a signature arg to a repeatable event, any existing periods become hidden in the UI (at present; perhaps that should change).

> This means that we'll have much bigger fish to fry than worrying about bugs with repeat detection.

Indeed.

---

So I think that we have answered our question here:
- Yes, it is perhaps conceivable that an event would have its args improperly registered so that no primary ("signature") arg was present, and to fix this a signature arg could be added; but
- No, adding a signature arg changes the nature of the event from repeatable to non-repeatable, since an event can only have one signature arg.

**TL;DR**: No.
